### PR TITLE
Basic Unit Tests

### PR DESCRIPTION
**Pull request by matwld**
Add basic unit tests, the coverage is about 50% of the code. Need to add more edge cases but would like review before getting deeper. Also show learnings on mocking in jest.

### CODE REVIEW

File: src/libs/tmdk/tests/deploy.test.ts [hunk 1]
```diff
@@ -1,26 +1,265 @@
...
 describe("testing deploy", () => {
-  beforeEach(mockExit.mockClear);
+  beforeEach(() => {
+    twinmakerMock.reset();
+    s3Mock.reset();
+    jest.resetAllMocks();
+    mockExit.mockClear();
+  });
 
   test("deploy_givenNoTMDKProject_expectError", async () => {
```
Comment (diehbria) [inline src/libs/tmdk/tests/deploy.test.ts]: prefer test names such as 'test('throws error when given directory that does not exist')' i.e. human readable test descriptions in the for mat of "does X when Y"

File: src/libs/tmdk/tests/deploy.test.ts [hunk 2]
```diff
@@ -1,26 +1,265 @@
...
 
-jest.mock("../src/lib/aws-clients");
+jest.mock("fs/promises", () => {
+  return { readFile: jest.fn() };
+});
+const twinmakerMock = mockClient(IoTTwinMakerClient);
+const s3Mock = mockClient(S3Client);
+const fakeTmdkDir = "/tmp/deploy-unit-tests";
 
 describe("testing deploy", () => {
```
Comment (diehbria) [inline src/libs/tmdk/tests/deploy.test.ts]: all tests are always testing, so no need to say that the test is testing

File: src/libs/tmdk/tests/deploy.test.ts [hunk 3]
```diff
@@ -1,26 +1,265 @@
...
+    expect(await handler(argv2)).toBe(0);
+    expect(twinmakerMock.commandCalls(CreateComponentTypeCommand).length).toBe(
+      0
+    );
+    expect(twinmakerMock.commandCalls(CreateEntityCommand).length).toBe(0);
+    expect(twinmakerMock.commandCalls(CreateSceneCommand).length).toBe(0);
+  });
+
+  test("deploy_given1ComponentType_expectSuccess", async () => {
+    fs.existsSync = jest.fn(() => {
```
Comment (diehbria) [inline src/libs/tmdk/tests/deploy.test.ts]: stylistic nit: () => true

File: src/libs/tmdk/tests/deploy.test.ts [hunk 4]
```diff
@@ -1,26 +1,265 @@
...
     } as Arguments<Options>;
-    await expect(handler(argv2)).rejects.toThrow(Error);
-    expect(mockExit).toHaveBeenCalledWith(1);
+    expect(await handler(argv2)).toBe(0);
+    expect(twinmakerMock.commandCalls(CreateComponentTypeCommand).length).toBe(
+      0
+    );
+    expect(twinmakerMock.commandCalls(CreateEntityCommand).length).toBe(0);
+    expect(twinmakerMock.commandCalls(CreateSceneCommand).length).toBe(1);
+    expect(
```
Comment (diehbria) [inline src/libs/tmdk/tests/deploy.test.ts]: there's a much cleaner way to assert on function called, using matchers such as `toHaveBeenCAlled() or `toHaveBeenLastCalledWith()`: https://jestjs.io/docs/mock-functions#custom-matchers

File: src/libs/tmdk/tests/deploy.test.ts [hunk 5]
```diff
@@ -1,26 +1,265 @@
...
+      twinmakerMock.commandCalls(CreateComponentTypeCommand)[0].args[0].input
+    ).toStrictEqual({
+      workspaceId: workspaceId,
+      ...componentType1Definition,
+    });
+    expect(twinmakerMock.commandCalls(CreateEntityCommand).length).toBe(0);
+    expect(twinmakerMock.commandCalls(CreateSceneCommand).length).toBe(0);
+  });
+
+  test("deploy_given1SceneAnd1Model_expectSuccess", async () => {
```
Comment (diehbria) [inline src/libs/tmdk/tests/deploy.test.ts]: better test name: `it('deploys successfully when there is one scene and one model')`

File: src/libs/tmdk/tests/init.test.ts [hunk 6]
```diff
@@ -1,22 +1,59 @@
+/* eslint @typescript-eslint/no-var-requires: "off" */
```
Comment (diehbria) [inline src/libs/tmdk/tests/init.test.ts]: if you have a very strong reason to disable a rule, better to disable it for a single line as opposed to the entire file

File: src/libs/tmdk/tests/deploy.test.ts (hunk 1 above)
Comment (matwld) [reply src/libs/tmdk/tests/deploy.test.ts]: Got it, will update

File: src/libs/tmdk/tests/deploy.test.ts (hunk 2 above)
Comment (matwld) [reply src/libs/tmdk/tests/deploy.test.ts]: Ok, for test suites how do we prefer to name them?

File: src/libs/tmdk/tests/deploy.test.ts (hunk 3 above)
Comment (matwld) [reply src/libs/tmdk/tests/deploy.test.ts]: Got it, did not know that would work. Thanks!

File: src/libs/tmdk/tests/deploy.test.ts (hunk 5 above)
Comment (matwld) [reply src/libs/tmdk/tests/deploy.test.ts]: Got it, will update

File: src/libs/tmdk/tests/deploy.test.ts (hunk 4 above)
Comment (matwld) [reply src/libs/tmdk/tests/deploy.test.ts]: I don't think this works for the aws mocking as I used the aws-sdk-mock package, not jest. I could not get the jest mock for s3 to work and for twinmaker it was unwieldy. Would be happy to discuss offline

File: src/libs/tmdk/tests/deploy.test.ts (hunk 2 above)
Comment (diehbria) [reply src/libs/tmdk/tests/deploy.test.ts]: Typically use `describe` block to logically group tests, for example, in this test suite: https://github.com/awslabs/iot-app-kit/blob/main/packages/core/src/common/dataFilters.spec.ts

each function being tested in a file that exports several, has a describe block.

For your use case, i would probably do something such as:


describe('deploys succesfully')
and

describe('fails to deploy')

as a way to categorize your test suites.

Since your deploy tests are contained in a deploy test file, it's already implied (and clear in the jest test runner results) that they are all related to deployment, so a single file level wide describe('deploys') adds no information though its a common pattern

File: src/libs/tmdk/tests/init.test.ts (hunk 6 above)
Comment (matwld) [reply src/libs/tmdk/tests/init.test.ts]: I removed the code that required the rule to be disabled

File: src/libs/tmdk/jest.config.js [hunk 7]
```diff
@@ -6,4 +6,12 @@ module.exports = {
   testEnvironment: "node",
   testRegex: "/tests/.*\\.(test|spec)?\\.(ts|tsx)$",
   moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
+  coverageThreshold: {
+    global: {
+      statements: 80,
+      branches: 80,
+      functions: 80,
+      lines: 80,
```
Comment (johnnyw-aws) [inline src/libs/tmdk/jest.config.js:14]: Is this too high for an alpha product? (maybe start with 50/60% if at all?)

Is there a way for us to emphasize integration test coverage instead of unit test coverage?

File: src/libs/tmdk/tests/deploy.test.ts [hunk 8]
```diff
@@ -3,24 +3,246 @@
...
+  } as Arguments<Options>;
+  expect(await handler(argv2)).toBe(0);
+  expect(twinmakerMock.commandCalls(CreateComponentTypeCommand).length).toBe(
+    0
+  );
+  expect(twinmakerMock.commandCalls(CreateEntityCommand).length).toBe(0);
+  expect(twinmakerMock.commandCalls(CreateSceneCommand).length).toBe(0);
+});
+
+test("deploys successfully when given a tmdk project with one component type", async () => {
```
Comment (johnnyw-aws) [inline src/libs/tmdk/tests/deploy.test.ts:103]: Does this duplicate coverage from the integ test?

If so, would remove + de-prioritize happy-case coverage in unit tests and focus on negative cases. Mainly because I'd say the overall project is still in an alpha state so fine-grained happy cases may just cause us to have to update more places if things in the codebase change

File: src/libs/tmdk/jest.config.js (hunk 7 above)
Comment (diehbria) [reply src/libs/tmdk/jest.config.js:14]: we want to have coverage threshold for every package - at a minimum, i want my team to be able to upgrade dependencies for security and package health reasons with 100% confidence nothing breaks. What code coverage is necessary to enable an away team to perform routine package updates without concern that something may regress?

File: src/libs/tmdk/tests/deploy.test.ts (hunk 8 above)
Comment (matwld) [reply src/libs/tmdk/tests/deploy.test.ts:103]: This does duplicate case from the functional test but I feel these tests are better for narrowing down specific cases where code fails. Functional test is more to verify that tmdk interacts with aws correctly. Also if code changes change the underlying use of aws api calls I think it is reasonable to expect developer to update the test cases as well.

Say for instance we change how deploy verifies whether a component type was created by using list component types after trying to create all of them instead of verifying each one in turn. While it will be additional work to change the test case to match this new behavior, without this fine grained test there is greater chance developer misses some edge case with their refactor.

File: src/libs/tmdk/jest.config.js (hunk 7 above)
Comment (matwld) [reply src/libs/tmdk/jest.config.js:14]: This is from appkit's coverage targets. The tests do not fail if we do not meet these targets, it just gives a warning. We can discuss offline about realistic coverage targets.

Comment (johnnyw-aws) [reply src/libs/tmdk/jest.config.js:14]: Thanks Brian, given that this tool is in an "early adoption / alpha phase" I'd like our acceptance testing to essentially be driven by correctness as defined through our end-to-end functional tests. Let's discuss offline on the coverage targets

**Review by johnnyw-aws — APPROVED** [review]
